### DISCUSSION TITLE
Issues indexing matplotlib

### DISCUSSION THREAD

**Discussion by tjdcs** [microsoft/pylance-release#4952] [Q&A]
It seems pylance is having troubling indexing matplotlib, package depth and `includeAllSymbols` have both been increased, but still having issues. 

For example:

```python
import matplotlib
```

Then while typing some code `matplotlib.axe` I expect to be able to autocomplete to the axes module but no such module can be found. this is needed for being able to type my code with object oriented matplotlib stye. I.E using the matplotlib.axes.Axes object and matplotlib.figure.Figure objects. 

Am I missing something or an extra extension needed?

matplotlib version 3.8.0
pylance version: v2023.10.21

**Comment by rchiodo** [comment]
This is happening because you're referencing sub packages. You have to import them first. 

```python
import matplotlib
import matplotlib.axes

matplotlib.axes. <-- this should work now
```

The symbol `matplotlib.axes` doesn't exist until you import it.

You can reproduce this with a simple folder setup like so:

```
root
    - test.py
    - subfolder 
       - __init__.py
       - subsubfolder
           - __init__.py
           - test_sub.py
```

```python
# contents of test.py
import subfolder
subfolder.subsubfolder <-- This is not a valid symbol yet, so it's not listed in the completions. 

```

The completion provider does not list things that would cause an error, so that's why `matplotlib.axes` doesn't complete.

**Reply by tjdcs** [reply]
Ah i see. Thank you for the aditional information. 

in other projects, I think I am getting autocompletion when simply typing the name of a symbol like `Axes` and a suggested import of `From matplotlib.axes` for example, I've included a screenshot of another symbol from a deeply nested package. Perhaps because these symbols are usually explicitly exported in __all__ for various sub packages?

![Screenshot 2023-10-11 at 5 12 09 PM](https://github.com/microsoft/pylance-release/assets/6669558/13266b05-e97e-4943-b11b-9685268caf50)


The use case I'm specifically dealing with is copying some code from an older file to a new refactor and needing to add the imports back in. usually I find the symbols that are missing, go to them with the cursor and then `cmd`+`.` to try to find them but this is not working with the `Axes` symbol (and many others that I would prefer to not have to search the documentation for every time)

**Reply by rchiodo** [reply]
Yeah, that seems like something we could fix. That XYZ_to_CCT in your screenshot is how auto import works. But it only works on first level items. Those without a '.' already. It feels like we should do the same for second level module imports too. They would cause an error unless you accepted the auto import.

**Reply by rchiodo** [reply]
I created this issue: https://github.com/microsoft/pylance-release/issues/4955
